### ISSUE TITLE
Some UIScrollDelegate Methods are not working

### ISSUE THREAD

**Issue by jalpizar-GAP** [fuzz-productions/FZAccordionTableView#13]
Hello,

First thanks for create this, currently I'm using your pod and I have to implement certain functionality after the user scroll up or down the table view, so I was trying to use the following UIScrollDelegate methods in my view controller (which contains a FZAccordionTableView):

`- (void)scrollViewWillEndDragging:(UIScrollView *)scrollView withVelocity:(CGPoint)velocity targetContentOffset:(inout CGPoint *)targetContentOffset NS_AVAILABLE_IOS(5_0);`
`// called on finger up if the user dragged. decelerate is true if it will continue moving afterwards`
`- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate;`
`- (void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView;` 

Basically they never get called, in order to fix this (the first method for example) I added this method inside your class `FZAccordionTableView.m`:
`- (void)scrollViewWillEndDragging:(UIScrollView *)scrollView withVelocity:(CGPoint)velocity targetContentOffset:(inout CGPoint *)targetContentOffset {
    [self.subclassDelegate scrollViewWillEndDragging:scrollView withVelocity:velocity targetContentOffset:targetContentOffset];
}`
In summary looks like FZAccordionTableView is not propagating the UIScrollDelegate methods.

**Comment by kgaidis** [comment]
Hi, thanks for the report! Just here to confirm that that is indeed the case and I will try to look for a flexible solution.

**Comment by jalpizar-GAP** [comment]
Great! Thanks!

**Comment by kgaidis** [comment]
Should be now fixed in 0.2.1, let me know if it works for you too!

For sake of knowledge:

First, the reason it's not optimal to just implement delegate calls and then forward them is because we MUST implement EVERY single delegate call. This is not very flexible: any newly added delgate calls in later iOS versions immediately make the library "broken." It's error prone: some delegate calls have default return values, overriding the delegate calls means we have to conform to those default return values which sometimes aren't documented. And maybe there could even be performance issues from making unnecessary message sends. This is actually how the initial implementation worked!

This leads to how the library works now:
`FZAccordionTableView` works by assigning itself as the `UITableViewDelegate / UITableViewDataSource` to the `UITableView`. Any calls that `FZAccordionTableView` does NOT implement, get forwarded to the delegate/datasource of the subclasser via message forwarding (https://www.mikeash.com/pyblog/friday-qa-2009-03-27-objective-c-message-forwarding.html). 

The reason some `UIScrollViewDelegate` calls work, and some don't is because `UITableView` seems to implement some of them! `scrollViewWillEndDragging:withVelocity:targetContentOffset:` is one of those. Message forwarding only works IF a message is sent and there is no one there to receive it. Well, `UITableView` gladly took that message for us.

To fix that, I just moved the delegate calls to a separate class `FZAccordionTableViewDelegate` proxy so `UITableView` private implementation doesn't interfere.

**Comment by pjryan93** [comment]
Thanks for taking the time to write out a detailed explanation.  I learned a lot and appreciate it!
